esphome/core/entity_base.h

```diff
@@ -46,7 +46,9 @@ class EntityBase {
   void set_icon(const std::string &name);
 
  protected:
-  virtual uint32_t hash_base() = 0;
+  /// The hash_base() function has been deprecated. It is kept in this
+  /// class for now, to prevent external components from not compiling.
+  virtual uint32_t hash_base() { return 0L; }
```

Comment:
Not sure how this works with overrides, but cant hurt right?
```suggestion
  ESPDEPRECATED("This method is deprecated. It is kept in this class for now, to prevent external components from not compiling.", "2022.6.0")
  virtual uint32_t hash_base() { return 0L; }
```

Reply:
I played with that too (including `helpers.h` and stating the deprecation like that), but I wasn't sure how to verify that the deprecated attribute was actually used.

Reply:
Ah wait, I got what I did wrong. I shouldn't use a semi-colon at the end, since this is an attribute. I see that I did add the `;` though, rendering the attribute ineffective. I'll push an update for this.
Just tried it locally, and now I do see the warnings as expected 👍

Reply:
Oh wow, but a complete build looks quite crappy now. Every time a component includes entity_base.h, the deprecation warning is triggered. For example:

```
src/esphome/components/binary_sensor/automation.cpp:117:1: warning: 'virtual uint32_t esphome::EntityBase::hash_base()' is deprecated: This method is deprecated. It is kept in this class for now, to prevent external components from not compiling. [-Wdeprecated-declarations]
 }  // namespace esphome
 ^
In file included from src/esphome/components/binary_sensor/binary_sensor.h:4:0,
                 from src/esphome/components/binary_sensor/automation.h:8,
                 from src/esphome/components/binary_sensor/automation.cpp:1:
src/esphome/core/entity_base.h:51:20: note: declared here
   virtual uint32_t hash_base() { return 0L; }
```

So this might not be a really useful change after all. Agreed?

Reply:
Looking at the documentation, a clean deprecation attribute is not feasible. In C++ 14, new attribute handling is introduced that does seem to do the trick, but that's of course not usable here.

Reply:
Interesting as we have them in other places and I don't think they got triggered. Oh well. Fine as is.

Reply:
> Interesting as we have them in other places and I don't think they got triggered. Oh well. Fine as is.

Possibly a difference between stand alone functions and class member functions.